### PR TITLE
feat(ios): use localhost in DEBUG mode

### PR DESCRIPTION
**Pull request by aleclarson**
This lets me use Vite in development

### CODE REVIEW

File: src/core/common.hh [hunk 1]
```diff
@@ -120,6 +120,10 @@
 #define SSC_VERSION_HASH ""
 #endif
 
+#ifndef PORT
+#define PORT 0
```
Comment (aleclarson) [inline src/core/common.hh:124]: You might want a better default here?

File: src/mobile/ios.mm [hunk 2]
```diff
@@ -223,12 +224,18 @@ - (BOOL) application: (UIApplication*)application didFinishLaunchingWithOptions:
...
   NSString* allowed = [[NSBundle mainBundle] resourcePath];
-  NSString* url = [allowed stringByAppendingPathComponent:@"ui/index.html"];
-
-  [self.webview
-		loadFileURL: [NSURL fileURLWithPath: url]
-    allowingReadAccessToURL: [NSURL fileURLWithPath: allowed]
-  ];
+  NSString* url = [NSURL fileURLWithPath:[allowed stringByAppendingPathComponent:@"ui/index.html"]];
+
+#if DEBUG
```
Comment (aleclarson) [inline src/mobile/ios.mm:229]: There might be cases where `localhost` isn't desired in DEBUG mode

File: src/core/common.hh (hunk 1 above)
Comment (heapwolf) [reply src/core/common.hh:124]: I remember reading that iOS has certain ports that are known not to run anything, iirc, 9777 is one, but I cant be sure, let me see if I cant find that information somewhere.

**Comment by aleclarson** [comment]
Oh it looks like we'll need a `--host` option as well. Or maybe we can infer the IP of the machine running the iOS simulator.

File: src/cli/cli.cc [hunk 3]
```diff
@@ -831,6 +832,18 @@ int main (const int argc, const char* argv[]) {
         }
       }
 
+      auto host = optionValue(arg, "--host");
+      if (host.size() > 0) {
+        devHost = host;
+      } else {
+        if (flagBuildForIOS || flagBuildForAndroid) {
+          auto r = exec("ifconfig | grep -w 'inet' | awk '!match($2, \"^127.\") {print $2; exit}'");
```
Comment (aleclarson) [inline src/cli/cli.cc:840]: For iOS and Android builds, I try to infer the LAN IP of the dev machine.

File: src/cli/cli.cc [hunk 4]
```diff
@@ -831,6 +832,18 @@ int main (const int argc, const char* argv[]) {
...
+          if (r.exitCode == 0) {
+            devHost = r.output;
+          }
```
Comment (aleclarson) [inline src/cli/cli.cc:843]: For iOS and Android builds, I try to infer the LAN IP of the dev machine.

Comment (aleclarson) [inline src/cli/cli.cc:843]: For iOS and Android builds, I try to infer the LAN IP of the dev machine.

Comment (aleclarson) [reply src/cli/cli.cc:843]: Not sure if that command works on anything but macOS though

Comment (heapwolf) [reply src/cli/cli.cc:843]: it works on macOS without the escaping of the double quotes. This worked for me

```bash
ifconfig | grep -w 'inet' | awk '!match($2, "^127.") {print $2; exit}'
```

File: src/core/common.hh (hunk 1 above)
Comment (heapwolf) [reply src/core/common.hh]: let's go with it for now, but I think that `0` could be a problem for that reason in particular, there are a lot of well known programs that can come online and expect a certain port

File: src/cli/cli.cc (hunk 4 above)
Comment (aleclarson) [reply src/cli/cli.cc:843]: Agreed. The escaping is for the `exec` call :)

File: src/mobile/ios.mm [hunk 5]
```diff
@@ -247,12 +249,17 @@ -           (BOOL) application: (UIApplication*) application
...
-  [self.webview
-		loadFileURL: [NSURL fileURLWithPath: url]
-    allowingReadAccessToURL: [NSURL fileURLWithPath: allowed]
-  ];
+  NSURL* url = [NSURL fileURLWithPath:[allowed stringByAppendingPathComponent:@"ui/index.html"]];
+
+#if DEBUG
+  url = [NSURL URLWithString: [NSString stringWithFormat:@"http://%@:%@/", _host, _port]];
+  NSURLRequest *request = [NSURLRequest requestWithURL:url];
+  [self.webview loadFileRequest:request
```
Comment (aleclarson) [inline src/mobile/ios.mm:257]:
```
/Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS16.0.sdk//System/Library/Frameworks/WebKit.framework/Headers/WKWebView.h:567:1: note: 'loadFileRequest:allowingReadAccessToURL:' has been marked as being introduced in iOS 15.0 here, but the deployment target is iOS 14.0.0
- (WKNavigation *)loadFileRequest:(NSURLRequest *)request allowingReadAccessToURL:(NSURL *)readAccessURL API_AVAILABLE(macos(12.0), ios(15.0));
```

File: src/mobile/ios.mm [hunk 6]
```diff
@@ -247,12 +249,17 @@ -           (BOOL) application: (UIApplication*) application
...
+      allowingReadAccessToURL:[NSURL fileURLWithPath:allowed]];
```
Comment (aleclarson) [inline src/mobile/ios.mm:258]:
```
/Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS16.0.sdk//System/Library/Frameworks/WebKit.framework/Headers/WKWebView.h:567:1: 
note: 'loadFileRequest:allowingReadAccessToURL:' has been marked as being introduced 
in iOS 15.0 here, but the deployment target is iOS 14.0.0
- (WKNavigation *)loadFileRequest:(NSURLRequest *)request 
    allowingReadAccessToURL:(NSURL *)readAccessURL 
    API_AVAILABLE(macos(12.0), ios(15.0));
```

File: src/init.cc [hunk 7]
```diff
@@ -15,4 +13,13 @@ namespace SSC {
...
   }
+
+  const char* getDevHost () {
+    static const char* host = STR_VALUE(HOST);
+    return host;
+  }
+
+  int getDevPort () {
+    return PORT;
+  }
```
Comment (aleclarson) [inline src/init.cc:24]: Moved references to `HOST` and `PORT` macros into `init.cc`

File: src/cli/templates.hh [hunk 8]
```diff
@@ -769,6 +769,10 @@ constexpr auto gXCodeProject = R"ASCII(// !$*UTF8*$!
         );
         LIBRARY_SEARCH_PATHS = "$(PROJECT_DIR)/lib";
         MARKETING_VERSION = 1.0;
+        OTHER_CFLAGS = (
+          "-DHOST={{host}}",
+          "-DPORT={{port}}",
+        );
```
Comment (aleclarson) [inline src/cli/templates.hh:775]: This is where the `HOST` and `PORT` macros are set for an iOS project

**Pull request merged**